packages/x-charts/src/models/plugin.ts

```diff
@@ -0,0 +1,27 @@
+import { ChartSeriesType, ExtremumGetter, Formatter } from './seriesType/config';
+import { AxisDefaultized } from './axis';
+import { DefaultizedSeriesType } from './seriesType';
+import { ZAxisDefaultized } from './z-axis';
+
+type ColorProcessor<T extends ChartSeriesType> = (
+  series: DefaultizedSeriesType<T>,
+  xAxis?: AxisDefaultized,
+  yAxis?: AxisDefaultized,
+  zAxis?: ZAxisDefaultized,
+) => (dataIndex: number) => string;
+
+export type ColorProcessorsConfig<T extends ChartSeriesType> = {
+  [Key in T]?: ColorProcessor<Key>;
+};
+
+export type ChartsPluginType<T extends ChartSeriesType> = {
+  seriesType: T;
+  seriesFormatter: Formatter<T>;
+  colorProcessor: ColorProcessor<T>;
+  xExtremumGetter?: ExtremumGetter<T>;
+  yExtremumGetter?: ExtremumGetter<T>;
+};
+
+export type ChartsPluginTypes<T extends ChartSeriesType> = {
+  [Key in T]: ChartsPluginType<Key>;
+}[T];
```

Comment:
This type seems useless to me, the result is simply `ChartsPluginType<T>`

Reply:
I did not find docs about that but TS does not react the same way to a generic and a union and union of generics

From what I understand, if you define `F<X> = {a: A<X>; b: B<X>}`

```ts
F<x|y> = {
  a: A<x|y>;
  b: B<x|y>;
}

F<x> | F<y> = { a: A<x>; b: B<x>; } | { a: A<y>; b: B<y>; }
```

If you use the generic for a single property it's the same. But as soon as you have two properties, it's not the same

Reply:
```tsx
export type ChartsPluginTypes<T extends ChartSeriesType> = {
  [Key in T]: ChartsPluginType<Key>;
}[T];
  ^--- This `[T]` 
```

Means you are indexing an object that is `Record<T, ChartsPluginType<T>>[T]`, which is exactly `ChartsPluginType<T>`

So in the end there is no union, it is just what seems an unnecessary step `{key: value}['key']`

Reply:
Oh that's new to me, the object assignment and subsequent access makes typescript calculate the discriminate union of the types instead of using the generic 🤔

Reply:
Ok, I found out why we needed `ChartsPluginTypes` and how to modify `ChartsPluginType` so it already trims down to the correct types without needing a different type. Why TS doesn't do this from the `<T extends ChartsSeriesType>` is beyond me though.

```tsx
export type ChartsPluginType<T> = T extends ChartSeriesType
  ? {
      seriesType: T;
      seriesFormatter: Formatter<T>;
      colorProcessor: ColorProcessor<T>;
      xExtremumGetter?: ExtremumGetter<T>;
      yExtremumGetter?: ExtremumGetter<T>;
    }
  : never;
```

you can see them in this: https://github.com/mui/mui-x/pull/13396